Feat(Mesh): Data plane proxy (#3892)
Kong/developer.konghq.com: app/mesh/data-plane-proxy.md

tags:
  - data-plane

related_resources:
  - text: 'Data plane on Kubernetes'
    url: '/mesh/data-plane-kubernetes/'
  - text: 'Data plane on Universal'
    url: '/mesh/data-plane-universal/'
  - text: Multi-zone deployment
    url: '/mesh/mesh-multizone-service-deployment/'
  - text: Zone Ingress
    url: /mesh/zone-ingress/
---

A data plane proxy (DPP), also known as a sidecar, is the part of {{site.mesh_product_name}} that runs next to each workload that is a member of the mesh.
A DPP is composed of the following components:
* A `Dataplane` entity defines the configuration of the DPP.
* A `kuma-dp` binary runs on each instance that is part of the mesh. This binary spawns the following subprocesses:
  * `Envoy` receives configuration from the control plane to manage traffic correctly.
  * `core-dns` resolves {{site.mesh_product_name}} specific DNS entries.

We have one instance of `kuma-dp` for every instance of every service.

## Concepts

The following sections define the core elements of the data plane proxy configuration.

### Inbound

An inbound defines the port exposed for each workload. It consists of:
* The port the workload listens on.
* A set of [tags](#tags).

> Most of the time a DPP exposes a single inbound. When a workload exposes multiple ports, multiple inbounds can be defined.

Reviewer: ```suggestion
A DPP typically exposes a single inbound. When a workload exposes multiple ports, multiple inbounds can be defined.
```

### Tags
Tags are a set of key-value pairs that are defined for each DPP inbound. These tags can serve the following purposes:
* Specifying the service this DPP inbound is part of.
* Adding metadata about the exposed service.

> * Allowing subsets of DPPs to be selected by these tags/

Reviewer: ```suggestion
* Allowing subsets of DPPs to be selected by these tags.
```

### Outbounds
An outbound allows the workload to consume a service in the mesh using a local port.
This is only useful when you're not using [transparent proxying](/mesh/transparent-proxying/). 

## `Dataplane` entity

The `Dataplane` entity consists of:

* The IP address used by other DPPs to connect to this DPP
* The inbounds configuration
* The outbounds configuration

A `Dataplane` entity must be present for each DPP. `Dataplane` entities are managed differently depending on the environment: 
* On [Kubernetes](/mesh/data-plane-kubernetes/), the control plane automatically generates the `Dataplane` entity. 
* On [Universal](/mesh/data-plane-universal/), the user defines the `Dataplane` entity. 
 
## Dynamic configuration of the data plane proxy 

When the DPP runs:
* The `kuma-dp` retrieves the Envoy startup configuration from the control plane.
* The `kuma-dp` process starts Envoy with this configuration.
* Envoy connects to the control plane using XDS and receives configuration updates when the state of the mesh changes.

The control plane uses policies and `Dataplane` entities to generate the DPP configuration. 

### Data plane proxy ports

> The `kuma-dp` process and its child process offer a number of services. These services need to listen to a few ports to provide their functionalities. 

Reviewer: This sentence feels a bit vague.
